plugins_src/primitives/wpc_thread.erl

```diff
@@ -0,0 +1,174 @@
+make_thread(Arg, _St) ->
+    ArgDict = dict:from_list(Arg),
+    Sections = dict:fetch(sections, ArgDict),
+    Pitch = dict:fetch(pitch, ArgDict),
+    TopRadius = dict:fetch(top_radius, ArgDict),
+    BotRadius = dict:fetch(bottom_radius, ArgDict),
+    CrestH = dict:fetch(crest_h, ArgDict),
+    Occurences = dict:fetch(occurences, ArgDict),
+    Dir = dict:fetch(direction, ArgDict),
+    Type = dict:fetch(thread_type, ArgDict),
+    Modify = [{dict:fetch(rot_x, ArgDict), dict:fetch(rot_y, ArgDict), dict:fetch(rot_z, ArgDict)},
+              {dict:fetch(mov_x, ArgDict), dict:fetch(mov_y, ArgDict), dict:fetch(mov_z, ArgDict)},
+              dict:fetch(ground, ArgDict)],
+
+    Height = Pitch*(Occurences+0.5),
+    Rows = Occurences+1,
+    make_thread(Type, Dir, Sections, Height, TopRadius, BotRadius, CrestH, Pitch, Rows, Modify).
+
+%%%
+%%% Thread
+%%%
+
+make_thread(Type, Dir, Sections, Height, TopRadius, BotRadius, CrestH, Pitch, Rows, [Rot, Mov, Ground]) ->
+    Vs1 = thread_verts(Type,Sections,TopRadius,BotRadius,CrestH,Pitch,Rows,Height),
+    Fs0 = thread_faces(Type,Sections,Rows),
+    {Vs0,Fs} = set_direction(Dir,Vs1,Fs0),
+    Vs = wings_shapes:transform_obj(Rot,Mov,Ground,Vs0),
+   {new_shape,thread(),Fs,Vs}.
+
+thread_verts(Type, Sections, TopRadius, BotRadius, CrestH, Pitch, Rows, Height) ->
+    RadRange = BotRadius-TopRadius,
+    RadInc = RadRange/(Rows-1),
+    Y = Height/2.0,
+    PitchInc = Pitch/2.0,
+    SubPitchInc = Pitch/Sections,
+    Delta = pi()*2/Sections,
+    Rings = lists:seq(Sections-1,0,-1),
+    lists:foldr(fun(Idx, Acc) ->
+                    YInc = Idx*Pitch,
+                    RInc = Idx*RadInc,
+                    Ring0 = ring_of_verts(Type, Rings, SubPitchInc, Delta, Y-YInc, TopRadius+RInc+CrestH),
+                    Ring1 = ring_of_verts(Type, Rings, SubPitchInc, Delta, Y-(YInc+PitchInc), TopRadius+RInc),
+                    Ring0++Ring1++Acc
+                end, [], lists:seq(0,Rows-1)).
+
+thread_faces(helicoid, N, Rows) ->
+    R0 = (Rows-1)*2,
+    R = (Rows*2-1)*N,
+    Top = lists:reverse(lists:seq(N-1,0,-1)),
+    Bottom = lists:seq(R+N-1,R, -1),
+    Ns = lists:seq(0, N-2),
+    Sides =
+        lists:foldl(fun(Idx, Acc) ->
+                        Sides = [[Idx*N + I, (Idx+1)*N + I,
+                                  (Idx+1)*N + ((I+1) rem N), Idx*N + ((Idx*N+I+1) rem N)] || I <- Ns],
+                        if Idx < (R0-1) ->
+                            Stitch = [[Idx*N + N-1, (Idx+1)*N + N-1, (Idx+2)*N + N, (Idx+2)*N]];
+                        Idx < R0 ->
+                            Stitch = [[Idx*N + N-1, (Idx+1)*N + N-1, (Idx+2)*N]];
+                        true ->
+                            Stitch = [[Idx*N + N-1, (Idx+1)*N + N-1, (Idx+1)*N]]
+                        end,
+                        Acc++Sides++Stitch
+                    end, [[0, N-1, N],[N-1, 2*N, N]], lists:seq(0,R0)),
+    [Top, Bottom | Sides];
```

Comment:
In general creating lists, just to loop over them with lists fold costs a lot of memory and performance,
instead of writing new functions that increment (or decrement) a counter.

So some lazy code here and in the next function, but I guess this isn't performance critical..